### PR TITLE
Search Block: add button only with expandable input

### LINKED ISSUE #31128: Search Block: Add icon-only option
## What problem does this address?

The search block comes with a number of easily configurable defaults:

search

Including:

- It's resizable
- Label can be toggled
- Button can be an icon button
- Button can be toggled off, shown outside, or shown inside

One nice addition would be the ability to show only the icon.

## What is your proposed solution?

Add an option to show just the search button:

Screenshot 2021-04-23 at 13 21 25

On focus the input field would then expand to allow search.

### PR DESCRIPTION
**Pull request by jffng**
## What?
Adds the Button (Icon) Only option to the search block. With this option selected, the behavior of the button changes so that when it's clicked, the search input field expands to the defined width.

## Why?
Solves #31128

## How?
If the revelant block attributes are present, a small client side script is enqueued to handle showing / hiding the search field. h/t @apeatling @aristath in #31719 for the starting point.

## Testing Instructions

1. Insert a search block to a post/page or in the site editor
2. Select the Button Only option from the toolbar
3. Select the button in the editor, verify the search field expands 
5. View the front end, click on it and verify focus is moved to the input. Type in a search term, click on the search button again and verify it performs the search.

### Testing Instructions for Keyboard
- Add a search block to a post or in the site editor, select the button only option
- Navigate to it using the keyboard / via tab. When the search block is focused, press enter / return. Verify the input field expands.
- Visit the front-end. Navigate to the search button. When it is focused, press enter / return. Verify the input expands and focus is moved to the input. Type in a search, press return, and verify.

## Screenshots or screencast 

**Editor**

https://github.com/WordPress/gutenberg/assets/5375500/4dca833f-2333-4bd9-b2fe-19b414c6b810 

**Front-end**

https://github.com/WordPress/gutenberg/assets/5375500/005c457a-b209-4778-b328-6eab5753a1a6

### CODE REVIEW

**Comment by jffng** [comment]
Hi @jasmussen — I was reading your most recent feedback on the previous PR: 

> This didn't quite work how I expected it — I expected that simply focusing the search button would make the search field appear, because the search "button" actually was the search field. Kind of [like this](https://codepen.io/k185/pen/PQajXE). I don't mind particularly the current behavior, and I can understand it might be necessary — so I would appreciate education on whether there might be some accessibility reason for why it should be this way. Is there a need for the ability to focus first the button, then invoke that with space, then transfer the focus to the input field?

It seems like having the focus jump to the input field by simply focusing on the button, would feel unexpected. Requiring the explicit user action to expand + focus the input, could feel less jarring. But I'm not sure, so wanted to ask again on this refreshed PR. 

Let me know what you think 🙏

**Comment by jasmussen** [comment]
First off, this is great. This has been a much needed feature for a long time. Trunk:

trunk

This branch:

Screenshot 2023-05-11 at 10 04 50

I think longer term we should revisit those dropdowns of options, the iconography isn't entirely solid, and there's arguably a better place in the inspector for some of these. I also think we can be better with the defaults, even keep the defaults smartly contextual so it's always collapsed into a labelles icon only version when inserted inside a container block.

But those are followups, and I think there's such value in this one that we should land it quickly. That said, in testing I'm seeing this work only in the editor, not on the frontend:

Screenshot 2023-05-11 at 10 05 37

> It seems like having the focus jump to the input field by simply focusing on the button, would feel unexpected. Requiring the explicit user action to expand + focus the input, could feel less jarring. But I'm not sure, so wanted to ask again on this refreshed PR.

I'm not sure the focus should jump anywhere. The way I've been thinking about this is that the button-only search button is a camouflaged input field. It's one and the same, and focusing it simply expands it so you can type. Kind of like one of these:

* https://codepen.io/jcoulterdesign/pen/pgwxJR
* https://codepen.io/CBeghin/pen/DbEmoB 
* https://codepen.io/lhenrique/pen/JvqGjj 
* https://codepen.io/AlbertFeynman/pen/BPvzWZ 
* https://codepen.io/omarsherifs/pen/KGLYYW
* https://codepen.io/aaroniker/pen/XyXzYp

None of those _look_ particularly great, but hopefully the principle is clear, that it's just a styled input field that looks like a search button when not focused. What do you think?

**Comment by jffng** [comment]
>  in testing I'm seeing this work only in the editor, not on the frontend:

Are you using the `npm run dev`? I think you may have to restart that process with this PR checked out, to trigger the block.json to be re-registered, since a new front-end script was introduced.

If that doesn't fix it, do you see any console errors?

**Comment by jasmussen** [comment]
npm run dev did it, apologies! Works decently well!

File: packages/block-library/src/search/block.json
Comment (mikachan) [inline packages/block-library/src/search/block.json]:
```suggestion
	"viewScript": "file:./view.min.js",
```

Should this be the `.min` version of the script? I see this is the version used in the [file block](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/file/block.json#L70) and the [navigation block](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/block.json#L137).

I'm also seeing a PHP notice on this branch, which the `.min` version of this script seems to fix:

```
Notice: Function register_block_script_handle was called <strong>incorrectly</strong>. The asset file (/build/block-library/blocks/search/view.asset.php) for the "viewScript" defined in "core/search" block definition is missing.
```

File: packages/block-library/src/search/index.php
Comment (mikachan) [inline packages/block-library/src/search/index.php]: Do we need to enqueue the `view.min.js` file here? This seems to be triggering multiple event listeners for me when I click the search button:

image

If I remove this line, `toggleSearchField` is only triggered once.

Although I know there's additional logic for its inclusion here (`button-only` and `expand-searchfield`), so maybe `view.min.js` shouldn't be included in block.json? I'm not sure which is the best, but I think it should only be included once. It's probably better if it's removed from block.json and kept here (in search/index.php), as then it's only included when the functionality is required.

**Review by mikachan — COMMENTED** [review]
This is working well for me!

I've left a couple of comments, mainly around the `view.js` file which seems to be included twice on the front end. I think once this is finalised, this is good to bring in.

File: packages/block-library/src/search/index.php
Comment (jffng) [reply packages/block-library/src/search/index.php]: Oh good call! I removed it from block.json [73176c1](https://github.com/WordPress/gutenberg/pull/50487/commits/73176c118e8098413b5880557930b4a57e1446a6). I thought it was needed to enqueue the file, but I guess that ends up loading it twice.

File: packages/block-library/src/search/block.json
Comment (jffng) [reply packages/block-library/src/search/block.json]: Should be resolved by [73176c1](https://github.com/WordPress/gutenberg/pull/50487/commits/73176c118e8098413b5880557930b4a57e1446a6).

**Review by mikachan — COMMENTED** [review]
> Oh good call! I removed it from block.json [73176c1](https://github.com/WordPress/gutenberg/pull/50487/commits/73176c118e8098413b5880557930b4a57e1446a6).

Nice, that's worked on my side!

One more thing, it looks like we might need to run `npm run fixtures:regenerate` to regenerate the test fixtures for the search block, so they include the new block attributes.

File: packages/block-library/src/search/edit.js
Comment (scruffian) [inline packages/block-library/src/search/edit.js]: Ideally we should list all the dependencies for this effect
```suggestion
	}, [ hasOnlyButton, isSelected, setAttributes ] );
```

File: packages/block-library/src/search/edit.js
Comment (scruffian) [inline packages/block-library/src/search/edit.js]:
```suggestion
	}, [ hasOnlyButton, isSelected, setAttributes, width ] );
```

File: packages/block-library/src/search/index.php
Comment (scruffian) [inline packages/block-library/src/search/index.php:82]: It would be nice to do this using the `WP_HTML_Tag_Processor`. There's an example of this in https://github.com/WordPress/gutenberg/pull/49212/files. Not necessary to get his merged, but a nice improvement.

File: packages/block-library/src/search/view.js
Comment (scruffian) [inline packages/block-library/src/search/view.js:1]: I wonder if this could use the new interactivity API cc @gziolo

File: packages/block-library/src/search/edit.js
Comment (jffng) [reply packages/block-library/src/search/edit.js]: [0cb5ef2](https://github.com/WordPress/gutenberg/pull/50487/commits/0cb5ef222e370ea31f0369abf46ad5067449af69)

File: packages/block-library/src/search/index.php
Comment (jffng) [reply packages/block-library/src/search/index.php:82]: Nice, happy to try that in a follow up.

File: packages/block-library/src/search/view.js
Comment (gziolo) [reply packages/block-library/src/search/view.js:1]: Yes, it definitely could. We will run the experiment for Interactivity API until everything is properly tested so it's fine to have two concurrent solutions for 2-3 Gutenberg plugin releases 👍🏻

**Comment by jffng** [comment]
Hi @jasmussen I pushed a change so that the input expands when the button is focused. How is it feeling to you?

**Comment by jasmussen** [comment]
Because the feature is so longed for and I want so much for it to land, I'm not sure I trust my ability to review this one, so CC: @WordPress/gutenberg-design for a broader view.

Notably a few things come to mind:

testing again

* On initial load, it animates from its expanded to its collapsed state
* If I focus the search button with tabbing, I have to first press enter to expand it, then I type my phrase, then I can press Enter. Would it be nice if focus expanded it automatically?
* Notably if I focus out of the box, it doesn't collapse. I have to actually click outside to collapse it. 

I'm assuming there's a reason that [the onfocus suggestions I suggested here](https://github.com/WordPress/gutenberg/pull/50487#issuecomment-1543544554) won't work for us?

**Comment by jffng** [comment]
> On initial load, it animates from its expanded to its collapsed state

Good catch, this should be addressed by [48b7ba3](https://github.com/WordPress/gutenberg/pull/50487/commits/48b7ba37e18fb6f4563f172fd08e40fedf13c13e) for the front-end. 

I can't seem to address this issue in the editor, but I don't know if the problem is specific to this PR.  I verified the block is initially rendered with the correct CSS classes, so theoretically the input should be hidden on initial render, but the transition is still happening noticeably. My hunch is that it's related to how the editor styles are loaded. Do you have any ideas?

> If I focus the search button with tabbing, I have to first press enter to expand it, then I type my phrase, then I can press Enter. Would it be nice if focus expanded it automatically?

This should be addressed by [05ba9a5](https://github.com/WordPress/gutenberg/pull/50487/commits/05ba9a5b83479fc4e5cde9f4b67281d5cd03a0cd), could you pull the latest changes and see if the behavior matches expectations? 

https://github.com/WordPress/gutenberg/assets/5375500/c28cd6b8-cf1a-4647-9d59-839d9878e0a6

> I'm assuming there's a reason that https://github.com/WordPress/gutenberg/pull/50487#issuecomment-1543544554 won't work for us?

No I think we can get there, I'm trying to match the behavior in those examples without altering the markup for this block significantly.

**Comment by jasmussen** [comment]
> No I think we can get there, I'm trying to match the behavior in those examples without altering the markup for this block significantly.

Ah of course, that makes sense. 

It's getting closer, though still seeing some issues.

* If I shift tab from the input field, it doesn't close. It only closes if I tab from the search button forward.
* If I click the input field after it was opened, it collapses.

**Review by vcanales — APPROVED** [review]
I took a look here, and it works correctly.

The latest comments by @jasmussen seem to have been addressed as well.

> If I shift tab from the input field, it doesn't close. It only closes if I tab from the search button forward.
> If I click the input field after it was opened, it collapses.

This LGTM

**Comment by jffng** [comment]
@jasmussen might you be able to give this another spin? Thank you for all the reviews already! 🙏

**Comment by jasmussen** [comment]
Took it for a spin, this is the behavior:

search

Notably on the frontend, I tab into the search box and it expands, and then I tab onwards to the search button. If I tab out, it collapses. 

If I shift tab from below the search box, I first tab into the button, and then backwards to the search field, which on focus expands. This part was a little surprising to me, because visually it appears as only a single button, so it would seem like a single tab element. 

I understand the challenge of keeping the markup as is, and it's an important feature to land. But I wonder if we should boost our confidence on the behavior here.

**Comment by richtabor** [comment]
> I think longer term we should revisit those dropdowns of options, the iconography isn't entirely solid, and there's arguably a better place in the inspector for some of these.

My initial thought was that this could be a block variation. One click to get the right combination of attributes that make and expandable search.

**Comment by jffng** [comment]
> Notably on the frontend, I tab into the search box and it expands, and then I tab onwards to the search button. 

Are you suggesting that tabbing onwards from the search box would skip the button and collapse the search box? Or the button is hidden entirely when the search box is expanded? 

Either way, I'm not sure these are good ideas, since the button is a standard affordance for both mouse and keyboard users to submit a form. To keep this affordance, the button needs to remain tab accessible and a distinct element from the input. 

> If I shift tab from below the search box, I first tab into the button, and then backwards to the search field, which on focus expands. This part was a little surprising to me, because visually it appears as only a single button, so it would seem like a single tab element.

I see what you mean. I'm not sure if we should address this. It goes back to my previous point that I think the button needs to remain a distinct element from the input that is tab accessible. As is, the button is semantic and associated with an action — in this case performing the search (and its associated function to expand it if it's hidden).

> I understand the challenge of keeping the markup as is

To clarify my earlier reasoning, it's not about keeping the markup as is just for the sake of it. The block uses standard html elements (an input and a button inside a form), and their current use is semantically meaningful. 

If we were to change the behavior to match an example closer to [this](https://codepen.io/AlbertFeynman/pen/BPvzWZ), it would require the addition of a non-semantic element (a div or a span that looks like a button but isn't one), and / or some CSS that would not be configurable by global styles elements API. 

>  I wonder if we should boost our confidence on the behavior here.

Yeah, how do you recommend going about this? Should we get additional input on the behavior? Clarify the requirements?

**Comment by jasmussen** [comment]
> Are you suggesting that tabbing onwards from the search box would skip the button and collapse the search box? Or the button is hidden entirely when the search box is expanded?

If we're keeping an input _and_ a separate button (as opposed to [the input-only options](https://github.com/WordPress/gutenberg/pull/50487#issuecomment-1543544554)), then that definitely seems a behavior to keep, I'm just describing it. 

No the nuance I'm observing here is that _visually_ there's only a button. But when I tab from prior elements towards the search box, the input expands from being invisible to being fully expanded. So I'm essentially tabbing into something I didn't know was there. 

That's the part where I'm unsure of the best practices — usually tab order should match the visual order, which in this case would set focus on the button directly. And this is the part to boost confidence in.

File: packages/block-library/src/search/view.js
Comment (mikachan) [inline packages/block-library/src/search/view.js:33]: We could also toggle a few additional attributes for accessibility purposes here.

When the search text input is hidden, we could add:
- `aria-hidden="true"` to the text input. It's worth noting that this prevents focus, but the element _is_ hidden. I'm not sure what the best practice would be here.
- `tabindex="-1"` also to the text input, so that the hidden input cannot be tabbed to
- `aria-expanded="false"` and `aria-controls="wp-block-search__input-[element-ID]"` to the search button (the element-ID would be the ID assigned when the search block is inserted). More info about [aria-expanded](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded).
- Possibly also consider adding an `aria-label="Expand/show search input"`, so that it's clear what the button will do

When the text input is visible:
- `aria-hidden="false"` on the text input. This allows focus again too, as well as being an accurate state.
- Remove `tabindex="-1"` from the text input
- `aria-expanded="true"` on the search button
- Remove `aria-label`, as the button text should suffice when it's acting as a search button

I think this would help make the experience of using a mouse vs using a keyboard more consistent.

**Review by mikachan — COMMENTED** [review]
I've given this another spin from an accessibility point of view, and I noticed there are probably a few attributes we should be using to describe this behaviour better. 

They may also help with this particular feedback from @jasmussen:

> usually tab order should match the visual order, which in this case would set focus on the button directly

If we stop the text input from being tabbable whilst it's hidden (as I've suggested by using `tabindex="-1"`), this means the only way to make the input visible is by hitting enter on the search button. Hopefully this also addresses this feedback:

> But when I tab from prior elements towards the search box, the input expands from being invisible to being fully expanded. So I'm essentially tabbing into something I didn't know was there.

I'm not sure what the best practice is here for these attributes in this context, but I think they're at least helping describe the visual behaviour better.

File: packages/block-library/src/search/view.js
Comment (jffng) [reply packages/block-library/src/search/view.js:33]: Nice suggestions! I applied them here: 3a5c81af6ce638b0c891f368b99e2098eecd4634

**Comment by jeryj** [comment]
I tested this out today, and, as someone who does enjoy interacting with sites via keyboard, the input auto-expanding when focusing toward the button was a bit jarring. I'd much prefer the experience of pressing `Enter` on the button to open the search, which would auto-focus the element for me.

The opening on focus experience only seems to impact keyboard users, and doesn't save them any tab stops from if the input was open by default, as you still have to press Tab twice to move past it. I'd recommend keeping the input hidden until the button is pressed explicitly with `Enter` or `Spacebar`.

File: packages/block-library/src/search/view.js
Comment (jeryj) [reply packages/block-library/src/search/view.js:33]: In the closed state, the button is responsible for displaying the search form. In the open state, the button is responsible for _submitting_ the form. This is an important distinction, as we're changing the behavior of the button, and need to communicate that. So, I think we need to modify a couple of *the properties in the open state*:

- Remove `aria-expanded`: It's no longer controlling the hidden input, and having `aria-expanded` set makes screen readers announce it as `Search, expanded, button`, which makes it seem like pressing it again should collapse it instead of perform a search.
- Remove `aria-controls` for the same reason. It's no longer controlling the element.

When the search input is hidden, I agree with @mikachan that the search button should have a label to communicate that it will open the search form.

Thanks for working on this @jffng!

Comment (jffng) [reply packages/block-library/src/search/view.js:33]: Thanks @jeryj — should we keep `aria-expanded` and `aria-controls` when the search input is hidden, to accurately describe the button behavior? Or is that taken care of by adding an `aria-label` to the button?

Comment (jffng) [reply packages/block-library/src/search/view.js:33]: I applied this feedback in 919117f6a58f7b3ec9be83904dcd8172ea3f0018

**Comment by ciampo** [comment]
Adding to @jeryj 's [comment](https://github.com/WordPress/gutenberg/pull/50487#issuecomment-1557970123), the current focus behavior feels unusual and a bit disorienting to me.

Here's an alternative that came to mind:
- by default, the input is _visually_ hidden (but accessible to assistive technology) and the button is visible. Both input and button represent tab stops as they would normally (no need for `tabindex` or `aria-hidden` attributes)
- when the input is focused, it becomes visible. The input would also stay visible while the button has focus, and (if we wanted) also while the input is not empty.

This way, both the input and the button would be always accessible to assistive technology, without altering the tab order, and while keeping the button to always be a form submit button.

**Comment by jffng** [comment]
Hi @ciampo thanks for weighing in. The alternative you suggest is how it worked it in [b028cbb](https://github.com/WordPress/gutenberg/pull/50487/commits/b028cbbb12801580812ef24138e866e3f327196e).

Tabbing to a hidden element also felt unexpected, and there's some uncertainty around whether an element that's visually hidden should be available in the tab order. So I revised the behavior to the current behavior — requiring a click or keypress to expand the search field.

Both directions have tradeoffs:

**button click or keypress required to expand the search field** | test
--- | ---
Adds more client side logic to handle the aria attributes.The big one with the current behavior is that the button's function sort of changes depending if the input is open or not |a sd

Feels unexpected

Which is worse?

**Comment by jffng** [comment]
Hi @ciampo thanks for weighing in. The alternative you suggest is how it worked it in [b028cbb](https://github.com/WordPress/gutenberg/pull/50487/commits/b028cbbb12801580812ef24138e866e3f327196e).

Tabbing to a hidden element also felt unexpected, and there's some uncertainty around whether an element that's visually hidden should be available in the tab order. So the current state of the PR requires a click or keypress to expand the search field.

I don't like that this one is a bit convoluted (the button purpose changes depending on the state of the search field, requires some additional client side js to handle aria attributes / tabindex), but I do think it provides a better UX for keyboard navigation.

File: packages/block-library/src/search/view.js
Comment (andrewhayward) [inline packages/block-library/src/search/view.js:32]: This causes problems if `searchButton` never had an `aria-label`. It will set the `aria-label` attribute to the string value "null", overriding any visible label the button might have. We should either check that `ariaLabel` has a value before doing this, or give `ariaLabel` a fallback value of "" above on line 13.

File: packages/block-library/src/search/view.js
Comment (andrewhayward) [inline packages/block-library/src/search/view.js]: It's not clear why we're removing `aria-label` here. If the button has a label, it's probably because it's an icon button. In which case, removing the attribute will leave the button nameless. Am I missing something?

File: packages/block-library/src/search/view.js
Comment (andrewhayward) [inline packages/block-library/src/search/view.js]: This event listener doesn't seem to do anything, I think because of the way `hideSearchField` checks to see if `searchField` is the active element.

**Comment by andrewhayward** [comment]
I have some comments, but first a newbie question – what is the thought process behind including this behaviour? Aside from it being a nice addition (#31128), have we validated that it's a useful and wanted bit of functionality, for both creators and consumers? I'm always a little wary of including things just because we _think_ something will be popular. Happy to be shown previous conversations around this.

Anyway, thoughts (in a bit of a brain dump – sorry!)...

When you click on a `<label>`, the browser is supposed to focus its related field. Unfortunately the `click` event listener on `document.body` obscures this behaviour here, by immediately removing focus from the field. Ideally this would be corrected.

The field remains hidden when it receives focus, potentially leaving the user in a confusing state. For most users this won't be an issue, because they can neither click on the label (see 1), nor tab to the input. But assistive technologies allow users to navigate in different ways, meaning it is possible to focus on the field, and then not be able to see it. The field should have a `focus` listener to make it visible, even if for most people it'll never be used.

It looks like the expectation is that if the search form loses focus, the field should collapse. But it doesn't expand when it gains focus. This feels quite unbalanced to me, as an interactive process. Not that I necessarily think it should expand and collapse based purely on focus – that could get quite jarring. I suppose what I'm thinking is that I'm not sure about the multi-use single-button approach, but that's more my gut than any evidence-based appraisal!

Finally, is it expected that if the field has a value (for example, on the results page, or if I just don't submit the search) it still collapses? This feels weird to me.

File: packages/block-library/src/search/view.js
Comment (jffng) [reply packages/block-library/src/search/view.js:32]: `searchButton` should always have an `aria-label`, which is defined [here](https://github.com/WordPress/gutenberg/pull/50487/files#diff-ef7ce82e804956a7d5c001fefbceabf1820872331c6282c9d359cac1e7390734R128-R130).

File: packages/block-library/src/search/view.js
Comment (jffng) [reply packages/block-library/src/search/view.js]: We remove `aria-label` here since the behavior of the button changes to its default behavior (`type="submit"`). Previously the label would read `Expand search button`. We could set the label to something else, but I thought this would add some unnecessary complexity since the button is performing its expected behavior in this state.

**Comment by jffng** [comment]
@andrewhayward thank you for the thoughts and feedback!

> have we validated that it's a useful and wanted bit of functionality, for both creators and consumers?

I've seen this pattern come up a lot from theme designers, and it was [prioritized here](https://github.com/orgs/WordPress/projects/96) as part of a larger effort to improve the Search Block.

> When you click on a <label>, the browser is supposed to focus its related field. Unfortunately the click event listener on document.body obscures this behaviour here, by immediately removing focus from the field. Ideally this would be corrected.

Nice, will address this 👍 EDIT: addressed in [7665720](https://github.com/WordPress/gutenberg/pull/50487/commits/7665720906561a8b71ba59d3a468cd75fa9a413c).

> (A) But assistive technologies allow users to navigate in different ways, meaning it is possible to focus on the field, and then not be able to see it. The field should have a focus listener to make it visible, even if for most people it'll never be used.

> (B) Not that I necessarily think it should expand and collapse based purely on focus – that could get quite jarring

I'm not sure I follow these pieces of feedback together — if the field has a focus listener to make it visible (A), how would it not expand on focus (B)?

>  is it expected that if the field has a value (for example, on the results page, or if I just don't submit the search) it still collapses?

Yes, that was the intention. It seemed related to [these guidelines](https://www.w3.org/TR/WCAG21/#content-on-hover-or-focus), so I thought there should be a way to collapse / dismiss the field, even if the field has a value.

File: packages/block-library/src/search/view.js
Comment (jffng) [reply packages/block-library/src/search/view.js]: You're right. I removed this event listener in [7665720](https://github.com/WordPress/gutenberg/pull/50487/commits/7665720906561a8b71ba59d3a468cd75fa9a413c) in since it wasn't doing anything, and also feels jarring to hide the field solely on focus loss.

File: packages/block-library/src/search/view.js
Comment (jffng) [reply packages/block-library/src/search/view.js]: I updated this so that aria-label is set to "Submit Search", instead of removing it: [fe7ff74](https://github.com/WordPress/gutenberg/pull/50487/commits/fe7ff74e97fe82fbd0ce3ddc3e57e3d44d2e6646)

**Comment by jffng** [comment]
I made a few more enhancements for keyboard and assistive technology users:

- `aria-label` and `type` attributes are updated to reflect the button's purpose (either to submit or expand the search)
- `escape` closes the form when the input OR button has focus.
- Clicking the label focuses the search field

Is there anything else that should be addressed in this PR? There's varying opinions on what UX feels good, so it would be good to get something in and iterate, assuming this is still a feature we want to land.

**Review by mikachan — COMMENTED** [review]
Thanks for addressing all the feedback! This is testing well for me:

- Tested using a mouse ✅
  - Clicking the label focuses the search field
- Tested using a keyboard ✅
  - Escape closes the form when the input OR button has focus
- Tested with VoiceOver on a Mac using the keyboard to navigate ✅
  - `aria-label` and `type` attributes reflect the button's purpose
  - Element descriptions work well
- Tested using a touch device ✅
  - Also works well on several different resolutions

I've noticed that the `aria-` and related attributes aren't updated in the editor when you interact with this block. I think it would be good to update these to reflect the different states, but I'm not sure it should block this PR. What do you think?

**Comment by jffng** [comment]
Thanks @mikachan !

> I've noticed that the aria- and related attributes aren't updated in the editor when you interact with this block. I think it would be good to update these to reflect the different states

Yes, though I'm not exactly sure what those attributes / labels need to be since their purpose differs in the editor context, e.g. the button doesn't submit the form in the expanded state. So yes, a follow up could be good there!

**Review by mikachan — APPROVED** [review]
> since their purpose differs in the editor context

Yeah, I was thinking this too, as I was trying to think of suggestions earlier but it's not an easy one.

With that in mind, I think this is ready to bring in. This is a great addition to the search block! 🎉

**Comment by jffng** [comment]
Thank you @mikachan! I opened up a follow up issue above to track these suggestions and will merge when the tests pass 💯

**Pull request merged**
